pubsub/fakepubsub/fake.go

```diff
@@ -0,0 +1,209 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fakepubsub provides an in-memory fake pubsub implementation.
+// This should not be used for production: it is intended for local
+// development.
+//
+// fakepubsub does not support any types for As.
+package fakepubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/google/go-cloud/pubsub/driver"
+)
+
+type topic struct {
+	mu        sync.Mutex
+	subs      []*subscription
+	nextAckID int
+	closed    bool
+}
+
+// CreateTopic creates a new topic.
+// Create subscribers for the topic before publishing.
+func CreateTopic() driver.Topic {
+	return &topic{}
+}
+
+// SendBatch sends the messages to all the subscriptions for this topic. It is an
+// error if the topic is closed or has no subscriptions.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// Check for closed or canceled before doing any work.
+	if t.closed {
+		return errors.New("fakepubsub: SendBatch: topic closed")
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	// Catch the common mistake of publishing before any subscriptions have been created.
+	if len(t.subs) == 0 {
+		return errors.New("fakepubsub: SendBatch: no subscriptions for topic")
+	}
+	// Associate ack IDs with messages here. It would be a bit better if each subscription's
+	// messages had their own ack IDs, so we could catch one subscription using ack IDs from another,
+	// but that would require copying all the messages.
+	for i, m := range ms {
+		m.AckID = t.nextAckID + i
+	}
+	t.nextAckID += len(ms)
+	for _, s := range t.subs {
+		s.add(ms)
+	}
+	return nil
+}
+
+// Close closes the topic. Subsequent calls to SendBatch will fail.
+func (t *topic) Close() error {
+	t.mu.Lock()
+	t.closed = true
+	t.mu.Unlock()
+	return nil
+}
+
+type subscription struct {
+	mu          sync.Mutex
+	topic       driver.Topic
+	ackDeadline time.Duration
+	msgs        map[driver.AckID]*message // all unacknowledged messages
+	ctx         context.Context           // for Close
+	cancel      func()
+}
+
+// CreateSubscription creates a new subscription for the given topic.
+// Unacknowledged messages will become available for redelivery after ackDeadline.
+// By default, real time (as determined by time.Now) is used. Set the variable Now
+// to a different function to mock time.
+func CreateSubscription(t driver.Topic, ackDeadline time.Duration) driver.Subscription {
```

Comment:
Or maybe Open? I was just rereading the design doc and that's what it uses.

I actually like Open because there's no confusion about whether we're actually creating a server-side resource or just a client-side handle for one.

Reply:
Definitely `Open` for the "real" implementations, so yeah maybe good to be consistent. You could argue that this impl really is creating the resource and so should be called `New`, but it's probably better to be consistent.